recovery_test.sh

```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+set -e
+set -x
+
+machine_id=$(cat /etc/machine-id)
+
+# Auto-detect first LUKS partition for testing
+test_device=$(lsblk -nlo NAME,TYPE,FSTYPE | awk '$2 == "part" && $3 == "crypto_LUKS" {print $1; exit}')
+if [ -z "$test_device" ]; then
+    echo "No LUKS partitions found, skipping recovery test"
+    exit 0
+fi
+
+test_partition=$(lsblk -nlo PARTLABEL /dev/$test_device)
+mapper_name=$(lsblk -nlo NAME /dev/$test_device | awk 'NR==2')
+
+echo "Testing recovery unlock for ${test_partition} (device: ${test_device}, mapper: ${mapper_name})..."
+
+# Get recovery key from Vault
+recovery_key=$(mangosctl sudo -- vault kv get -field=key "secrets/mangos/recovery-keys/${machine_id}/${test_partition}")
+
+# Find TPM keyslot number
+tpm_slot=$(cryptsetup luksDump /dev/$test_device | \
+    awk '/Tokens:/,/Keyslots:/ {if (/systemd-tpm2/) found=1; if (found && /^  [0-9]+:/) {print $1; exit}}' | \
+    tr -d ':')
+
+echo "Removing TPM keyslot ${tpm_slot} (simulating TPM failure)..."
+echo -n "$recovery_key" | cryptsetup luksKillSlot /dev/$test_device "$tpm_slot" --key-file=-
+
+# Get mount point for this partition
+mount_point=$(findmnt -n -o TARGET /dev/mapper/$mapper_name)
+
+# Unmount and close
+if [ -n "$mount_point" ]; then
+    systemctl stop $(systemd-escape -p --suffix=mount "$mount_point")
+fi
+cryptsetup close $mapper_name
+
+# THE CRITICAL TEST: Unlock with recovery key
+echo "Unlocking with recovery key..."
+echo -n "$recovery_key" | cryptsetup open /dev/$test_device $mapper_name --key-file=-
```

Comment:
This is equivalent to `systemd-cryptsetup attach $mapper_name $test_device $keyfile`. I'm not entirely sure if it'll let you pass it on stdin like you did.